src/transport_ofi.h

```diff
@@ -658,6 +658,139 @@ void shmem_transport_put_nb(shmem_transport_ctx_t* ctx, void *target, const void
+                                     };
+        const struct fi_rma_iov rma_iov = {
+                                            .addr = (uint64_t) addr,
+                                            .len = len,
+                                            .key = key
+                                          };
+        const struct fi_msg_rma msg = {
+                                        .msg_iov = &msg_iov,
+                                        .desc = NULL,
+                                        .iov_count = 1,
+                                        .addr = GET_DEST(dst),
+                                        .rma_iov = &rma_iov,
+                                        .rma_iov_count = 1,
+                                        .context = src_buf,
+                                        .data = 0
+                                      };
+
+        do {
+            ret = fi_writemsg(ctx->ep, &msg, FI_DELIVERY_COMPLETE);
+        } while (try_again(ctx, ret, &polled));
+
+        SHMEM_TRANSPORT_OFI_CTX_UNLOCK(ctx);
+    } else {
+        uint8_t *frag_source = (uint8_t *) source;
+        uint64_t frag_target = (uint64_t) addr;
+        size_t frag_len = len;
+
+        struct iovec msg_iov = {
+                                 .iov_base = frag_source,
+                                 .iov_len = frag_len
+                               };
+        struct fi_rma_iov rma_iov = {
+                                      .addr = frag_target,
+                                      .len = frag_len,
+                                      .key = key
+                                    };
+        struct fi_msg_rma msg = {
+                                  .msg_iov = &msg_iov,
+                                  .desc = NULL,
+                                  .iov_count = 1,
+                                  .addr = GET_DEST(dst),
+                                  .rma_iov = &rma_iov,
+                                  .rma_iov_count = 1,
+                                  .context = frag_source,
+                                  .data = 0
+                                };
+
+        SHMEM_TRANSPORT_OFI_CTX_LOCK(ctx);
+        while (frag_source < (((uint8_t *) source) + len)) {
+            frag_len = MIN(shmem_transport_ofi_max_msg_size, 
+                          (size_t) (((uint8_t *) source) + len - frag_source));
+            polled = 0;
+
+            msg_iov.iov_base = frag_source;
+            msg_iov.iov_len = frag_len;
+
+            rma_iov.addr = frag_target;
+            rma_iov.len = frag_len;
+
+            msg.msg_iov = &msg_iov;
+            msg.rma_iov = &rma_iov;
+            msg.context = frag_source;
+
+            SHMEM_TRANSPORT_OFI_CNTR_INC(&ctx->pending_put_cntr);
+
+            do {
+                ret = fi_writemsg(ctx->ep, &msg, FI_DELIVERY_COMPLETE);
+            } while (try_again(ctx, ret, &polled));
+
+            frag_source += frag_len;
+            frag_target += frag_len;
+        }
```

Comment:
Just making an observation that when fragmentation occurs, the signal will be signaling completion of multiple prior puts.  I re-read the `FI_FENCE` text and I think this is ok.  It might not hurt to toss a comment somewhere to the effect that multiple prior operations may be fenced by the signal put when fragmentation occurs.

Reply:
Added.